src/test/java/com/github/firmwehr/gentle/testutil/EqualityChecker.java

```diff
@@ -62,13 +72,56 @@ private boolean isEqualAttribute(MethodHandle handle, T actual) {
 			// if both are null, we consider them equal, if only one is null, we consider them not equal
 			return expectedAttribute == actualAttribute;
 		}
+		if (this.deepScanCollections && expectedAttribute instanceof Collection a &&
+			actualAttribute instanceof Collection b) {
+			return collectionsEqual(a, b);
+		}
+		// we need to handle Optional separately
+		if (this.deepScanRecords && expectedAttribute instanceof Optional<?> eO &&
+			actualAttribute instanceof Optional<?> aO) {
+			if (eO.isEmpty()) {
+				return aO.isEmpty();
+			}
+			Object nonOptionalExpectedAttribute = eO.get();
+			if (!nonOptionalExpectedAttribute.getClass().isRecord()) {
+				return nonOptionalExpectedAttribute.equals(aO.orElse(null));
+			}
+			return new EqualityChecker<>(nonOptionalExpectedAttribute, this.exceptTypes, true,
+				this.deepScanCollections).isEqual(aO.orElse(null));
+		}
 		// if we don't need to deeply scan records, we can just use normal equals here
-		if (!this.deep ||
+		if (!this.deepScanRecords ||
 			!(expectedAttribute.getClass().isRecord() && expectedAttribute.getClass() == actualAttribute.getClass())) {
 			return expectedAttribute.equals(actualAttribute);
 		}
 		// otherwise, we just run a check again on the attribute
-		return new EqualityChecker<>(expectedAttribute, this.exceptTypes, true).isEqual(actualAttribute);
+		return new EqualityChecker<>(expectedAttribute, this.exceptTypes, true, this.deepScanCollections).isEqual(
+			actualAttribute);
+	}
+
+	private boolean collectionsEqual(Collection<?> a, Collection<?> b) {
+		if (a == b) {
+			return true;
+		}
+		if (a.size() != b.size()) {
+			return false;
+		}
+		//noinspection SwitchStatementWithTooFewBranches
+		return switch (a) {
+			case List<?> list && b instanceof List<?> bList -> listsEqual(list, bList);
```

Comment:
Why do you need Lists here?

Reply:
The equality contract for Sets is a little bit more complex and can't be handled correctly the same way as lists can be handled. We need to add an implementation for that if we want to cover Sets too.